### PR TITLE
Feature/sublime plugin

### PR DESCRIPTION
**Pull request by Derfies**
Basic implementation of the plugin for sublime.

### CODE REVIEW

**Comment by donkirkby** [comment]
Sorry about the broken tests, there must have been a tox update in the last couple of weeks. I fixed the Travis script, and merged it into your branch.

I'm still having the same problem I mentioned before: when I follow the instructions in your README, I see the new menu items, but they're all disabled.

Here's what I see in my Sublime console:

```
startup, version: 3207 linux x64 channel: stable
executable: /snap/sublime-text/58/opt/sublime_text/sublime_text
working dir: /
packages path: /home/don/.config/sublime-text-3/Packages
state path: /home/don/.config/sublime-text-3/Local
zip path: /snap/sublime-text/58/opt/sublime_text/Packages
zip path: /home/don/.config/sublime-text-3/Installed Packages
ignored_packages: ["Vintage"]
pre session restore time: 0.0889609
startup time: 0.195765
first paint time: 0.212906
reloading plugin Default.arithmetic
reloading plugin Default.auto_indent_tag
reloading plugin Default.block
reloading plugin Default.colors
reloading plugin Default.comment
reloading plugin Default.convert_color_scheme
reloading plugin Default.convert_syntax
reloading plugin Default.copy_path
reloading plugin Default.detect_indentation
reloading plugin Default.echo
reloading plugin Default.exec
reloading plugin Default.fold
reloading plugin Default.font
reloading plugin Default.goto_line
reloading plugin Default.history_list
reloading plugin Default.indentation
reloading plugin Default.install_package_control
reloading plugin Default.kill_ring
reloading plugin Default.mark
reloading plugin Default.new_templates
reloading plugin Default.open_context_url
reloading plugin Default.open_in_browser
reloading plugin Default.pane
reloading plugin Default.paragraph
reloading plugin Default.paste_from_history
reloading plugin Default.profile
reloading plugin Default.quick_panel
reloading plugin Default.rename
reloading plugin Default.run_syntax_tests
reloading plugin Default.save_on_focus_lost
reloading plugin Default.scroll
reloading plugin Default.set_unsaved_view_name
reloading plugin Default.settings
reloading plugin Default.show_scope_name
reloading plugin Default.side_bar
reloading plugin Default.sort
reloading plugin Default.switch_file
reloading plugin Default.symbol
reloading plugin Default.transform
reloading plugin Default.transpose
reloading plugin Default.ui
reloading plugin CSS.css_completions
reloading plugin Diff.diff
reloading plugin HTML.encode_html_entities
reloading plugin HTML.html_completions
reloading plugin ShellScript.ShellScript
plugins loaded
```

How does that compare to what you see? Do you see something like "reloading plugin PythonLiveCoding"? Can you suggest any other ways for me to debug the problem in my environment?

**Comment by Derfies** [comment]
Short of getting a Linux box up on my end, or binary chopping until the code works, I can’t recommend anything unfortunately. I’ll paste full stdout here later but yes - I just get “reloading PythonLiveCoding”. Nothing more than that. All I know is that sublime does snake_case CamelCase mangling Unser the hood to map command as they’re written in the menu to classes in the plugin code. If that mapping fails or an exception is thrown then menus are greyed out, but it’s only in the latter case that you’ll see anything in the log.

I can supply a plugin minimum working example to see if that runs - if it does that gives us something to work towards but if it doesn’t then I suppose we have to complain in the sublime message boards that their Linux build is broken.

**Comment by Derfies** [comment]
Minimal pluging example:

[Test.zip](https://github.com/donkirkby/live-py-plugin/files/3218718/Test.zip)

File: sublime/pythonLiveCoding.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    settings = sublime.load_settings('PythonLiveCoding.sublime-settings')
+    py_path = settings.get('python_executable')
+    tracer_path = settings.get('code_tracer')
+    args = [py_path, tracer_path]
+    if pargs is not None:
+        args.extend(pargs)
+    args.append('-')
+
+    # Startup info so we don't open a new command prompt.
+    startupinfo = subprocess.STARTUPINFO()
```
Comment (donkirkby) [inline sublime/pythonLiveCoding.py]: According to the [Python docs]:

>The STARTUPINFO class and following constants are only available on Windows.

Would you like me to add some commits to make this work on GNU/Linux?

[Python docs]: https://docs.python.org/3/library/subprocess.html#windows-popen-helpers

Comment (Derfies) [reply sublime/pythonLiveCoding.py]: Fixed as per https://stackoverflow.com/questions/1016384/cross-platform-subprocess-with-hidden-window

**Comment by donkirkby** [comment]
Congratulations! It now works for me on Ubuntu.

File: sublime/pythonLiveCoding.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    """
+
+    pending = 0
+
+    @classmethod
+    def is_applicable(cls, settings):
+        return settings.has(OUTPUT_VIEW_ID)
+
+    def on_timeout_async(self):
+        self.pending = self.pending - 1
```
[the remark this answers is not in the document]
Comment (Derfies) [reply sublime/pythonLiveCoding.py]: Good call. Will make the change.

File: sublime/pythonLiveCoding.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        output_view.set_scratch(True)
+        output_view.settings().set('word_wrap', False)
+        output_view.set_name('live coding output')
+
+        input_view.settings().set('word_wrap', False)
+        input_view.settings().set(OUTPUT_VIEW_ID, output_view.id())
+        input_view.settings().set(INPUT_VIEW_TRACER_ARGS, self.TRACER_ARGS)
+
+        contents = input_view.substr(sublime.Region(0, input_view.size()))
+        update_view(output_view, *trace_code(contents, self.TRACER_ARGS))
```
[the remark this answers is not in the document]
Comment (Derfies) [reply sublime/pythonLiveCoding.py]: So just merge the two functions? Sure.

File: sublime/pythonLiveCoding.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+        stdin=subprocess.PIPE, 
+        stdout=subprocess.PIPE, 
+        stderr=subprocess.PIPE, 
+        startupinfo=startupinfo,
+        universal_newlines=True
+    )
+    return proc.communicate(input=input)
+
+
+def update_view(output_view, out, err):
```
[the remark this answers is not in the document]
Comment (Derfies) [reply sublime/pythonLiveCoding.py]: No worries. What was the error, out of interest?

**Comment by Derfies** [comment]
Awesome you got it to work!! What was the issue?

Regarding remaining work:
- I’ve requested an onscrollpositonhanged onnthe sublime message boards but there’s no movement there. I’m not going to hold my breath. I’ll have to add an onidle callback which isn’t going to be pretty - but it should work at least. 
- not sure what the best way of implementing turtle graphics is. If I’m understanding your code correctly, you output the turtle commands and then redraw that inside the plugin. If there’s no raw drawing commands available I’m not sure how to create the image (am I way off base here...??)

File: sublime/pythonLiveCoding.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (donkirkby) [reply sublime/pythonLiveCoding.py]: I had the wrong path to `code_tracer.py`. A dumb mistake that was easy to fix, once I saw the error output.

File: sublime/pythonLiveCoding.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (donkirkby) [reply sublime/pythonLiveCoding.py]: I meant calling `trace_code()` inside `update_view()` so you wouldn't have to pass the `stdout` and `stderr` text around as much. Merging would also work, so do whichever you prefer.

**Comment by donkirkby** [comment]
Sounds like a timer is a good approach to start with. Maybe we can come up with something better as we learn more about Sublime's internals.

It looks like drawing is harder than I expected. The only approach I can come up with is using a phantom like you did with the `create_image` command. The phantoms use [minihtml], but it only supports PNG, JPG, and GIF. No SVG or canvas support. I guess we could theoretically create a PNG image from the turtle commands, but that's way more work than I'm willing to tackle right now.

In summary: polish off the last few comments I made, add the timer for scrolling, and we're ready to merge!

[minihtml]: https://www.sublimetext.com/docs/3/minihtml.html

File: sublime/pythonLiveCoding.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
+import os
```
Comment (donkirkby) [inline sublime/pythonLiveCoding.py]: One more thing: could you please change the file name to `python_live_coding.py`? You'll probably need to change the package folder name and settings file to match.

Comment (Derfies) [reply sublime/pythonLiveCoding.py]: Can do. I think sublime plugin's convention is upper-leading camel case however - I'm not sure if there's a technical requirement for this to do with name mangling / mapping but that's how I've seen sublime plugins done in the past. Happy to make the change though.

File: sublime/pythonLiveCoding.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Derfies) [reply sublime/pythonLiveCoding.py]: Just looking at this again. I think the original example (http://www.sublimetext.com/docs/plugin-examples) is set up this way because you can't veto a callback once the timeout has been set. This means that the only way to know which is the most recent callback is to use a counter, some uuid, or look at milliseconds against the epoch or similar. If it's a bool there's no way to know how many keystrokes are still queued.

[the remark this answers is not in the document]
Comment (donkirkby) [reply sublime/pythonLiveCoding.py]: Thanks for the explanation. Looks good to me.

File: sublime/pythonLiveCoding.py (hunk 5 above)
Comment (donkirkby) [reply sublime/pythonLiveCoding.py]: I just tried it here, and `python_live_coding.py` works fine.

File: sublime/pythonLiveCoding.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Derfies) [reply sublime/pythonLiveCoding.py]: Ok I'm now calling one from the other. The functions sigs are a lot cleaner and it removes code from other places in the plugin so that's a win.

File: sublime/pythonLiveCoding.py (hunk 5 above)
Comment (Derfies) [reply sublime/pythonLiveCoding.py]: Ok, I've implemented an on_idle callback. It's not the prettiest thing in the world, but it does the trick.

**Comment by Derfies** [comment]
Ok, I've implemented an on_idle callback. It's not the prettiest thing in the world, but it keeps the views in sync. Aside from the module name, was there anything else outstanding?

**Comment by donkirkby** [comment]
The scrolling works fine. Just the name change, and then we're ready to merge.

Thanks for the great contributions!

**Comment by Derfies** [comment]
Done!

**Comment by donkirkby** [comment]
This pull request was for issue #175.
